src/main/java/com/openlattice/chronicle/services/message/MessageService.kt

```diff
@@ -0,0 +1,113 @@
+package com.openlattice.chronicle.services.message
+
+import com.google.common.collect.ArrayListMultimap
+import com.google.common.collect.ImmutableSet
+import com.google.common.collect.ListMultimap
+import com.openlattice.chronicle.constants.EdmConstants
+import com.openlattice.chronicle.data.MessageDetails
+import com.openlattice.chronicle.data.MessageOutcome
+import com.openlattice.chronicle.services.ApiCacheManager
+import com.openlattice.chronicle.services.edm.EdmCacheManager
+import com.openlattice.chronicle.services.enrollment.EnrollmentManager
+import com.openlattice.chronicle.services.entitysets.EntitySetIdsManager
+import com.openlattice.chronicle.services.twilio.TwilioManager
+import com.openlattice.client.ApiClient
+import com.openlattice.data.DataAssociation
+import com.openlattice.data.DataGraph
+import org.slf4j.LoggerFactory
+import java.util.*
+import java.util.concurrent.ExecutionException
+
+/**
+ * @author toddbergman &lt;todd@openlattice.com&gt;
+ */
+open class MessageService(
+        private val apiCacheManager: ApiCacheManager,
+        private val edmCacheManager: EdmCacheManager,
+        private val enrollmentManager: EnrollmentManager,
+        private val entitySetIdsManager: EntitySetIdsManager,
+        private val twilioManager: TwilioManager
+) : MessageManager {
+
+    companion object {
+        protected val logger = LoggerFactory.getLogger(MessageService::class.java)
+    }
+
+    override fun sendMessages(organizationId: UUID, messageDetailsList: List<MessageDetails>) {
+        val outcomes = twilioManager.sendMessages(messageDetailsList)
+        recordMessagesSent(organizationId, outcomes)
+    }
+
+    @Throws(ExecutionException::class)
+    fun recordMessagesSent(organizationId: UUID?, messageOutcomes: List<MessageOutcome>) {
+        val apiClient = apiCacheManager.prodApiClientCache[ApiClient::class.java]
+        val dataApi = apiClient.dataApi
+        val coreAppConfig = entitySetIdsManager.getChronicleAppConfig(organizationId)
+
+        // entity set ids
+        val messageESID = coreAppConfig.messagesEntitySetId
+        val sentToESID = coreAppConfig.sentToEntitySetId
+        val participantsESID = coreAppConfig.participantEntitySetId
+
+        val entities: ListMultimap<UUID, Map<UUID, Set<Any>>> = ArrayListMultimap.create();
+        val associations: ListMultimap<UUID, DataAssociation> = ArrayListMultimap.create()
+
+        messageOutcomes.forEachIndexed { index, messageOutcome ->
+            val participantEKID =
+                    enrollmentManager.getParticipantEntityKeyId(organizationId, messageOutcome.studyId, messageOutcome.participantId)
+
+            val messageEntity = mapOf(
+                edmCacheManager.getPropertyTypeId(EdmConstants.TYPE_FQN) to ImmutableSet.of<Any>(messageOutcome.messageType),
+                edmCacheManager.getPropertyTypeId(EdmConstants.OL_ID_FQN) to ImmutableSet.of<Any>(messageOutcome.sid),
```

Comment:
Generally speaking, this should be unique across entities because `ol.id` is the primary key. Isn't `sid` going to be the same every time?

Reply:
I don't think so. they will just all start with SM - indicating that they are SMS. They are still unique identifiers within twilio.

Reply:
you can look at this entity set if you want to get an idea of what sid looks like.
https://openlattice.com/orgs/#/orgs/e1aec603-1171-4c6b-a00e-b367dac07c9d/dataSets/51097efb-1647-476b-80c7-b8c31c168bd7/data